### ISSUE TITLE
[Performance] StableDiffusion on Mult-GPU

### ISSUE THREAD

**Issue by innat** [keras-team/keras-cv#1455]
## Issue

I run stable diffusion model (v1, v2) on single gpu and mult-gpu to observe the execution performance of the model. And get some possible issue. 

- For mult-gpu (2x Tesla-T4), the stable-diffusion v2 takes around ~ **28** sec to generate **3** images of size **512 x 512**, with step size **50**. 
- For single-gpu (1x Tesla-T4), **same-as-previous**.

Is it expected?  

Also, for mult-gpu (2x T4), I'm using [kaggle-env](https://www.kaggle.com/code). I noticed that both gpu's gpu-ram is full but only one gpu is being used.

<img src="https://user-images.githubusercontent.com/17668390/222331911-c87ea941-e7e2-49e1-97e6-d7b639258eb2.png"  width="30%" height="10%">

## Reproduce

[Colab gist](https://colab.research.google.com/drive/1bRqHb14TRziEcJqWZqQVuC_2HPw3c_bq?usp=sharing).  [Data-set.](https://drive.google.com/file/d/13zEZBjYXnQgOt2Q5niiiPBzwk4o-JhfE/view?usp=sharing)

cc. @LukeWood @ianstenbit @bhack 

**others**

- https://www.kaggle.com/competitions/stable-diffusion-image-to-prompts/overview

**Comment by innat** [comment]
Okay, I think I need to distribute the dataset (here which is prompt). 

```python
dataset = tf.data.Dataset.from_tensor_slices(df_train.sample(20).Prompt.values)
dataset = dataset.batch(12)
dataset = dataset.prefetch(tf.data.AUTOTUNE)
dataset = strategy.experimental_distribute_dataset(dataset)
```
Follow-up observaiton, does the current SD take `tf.data` as input?

**Comment by innat** [comment]
@miguelCalado I found your reported issue [here](https://github.com/divamgupta/stable-diffusion-tensorflow/issues/26#issuecomment-1259157338), and I think we're in the same page. Did you find any workaround?

**Comment by innat** [comment]
@miguelCalado Thanks for the hint. I tried as you suggested but unfotunately didn't see any change. Does it work on your case (ptp-tf)?

**Comment by miguelCalado** [comment]
Yes, I can say that in my case it worked... Maybe you could share a notebook to see if I can spot any errors?

**Comment by innat** [comment]
Here is the [gist](https://colab.research.google.com/drive/186dzamjNaH4YlUnHoNFEwVIoestLCKdD?usp=sharing).   
I repalced `predict_on_batch` with `predict` as you suggested, [modified script](https://github.com/innat/keras-cv/blob/rm_pb/keras_cv/models/stable_diffusion/stable_diffusion.py). You may need to run it mult-gpu devices (gcp or use kaggle). Sample [data](https://github.com/keras-team/keras-cv/issues/1455#issue-1606095073).

**Comment by miguelCalado** [comment]
I think the issue is that for the first time you run inference, the models are loaded and downloaded. So doing this, for example:
```python
image_gen_model = kcv_models.StableDiffusionV2(
         512, 512, jit_compile=True
    )
```
Doesn't load the model.

The solution is to:
```python
with strategy.scope():
    image_gen_model = kcv_models.StableDiffusionV2(
         512, 512, jit_compile=True
    )
    # to warm-up
    _ = generate_image(
            "A"*77, image_gen_model, image_per_prompt=2
        )
```

This way you get both GPUs working as demonstrated in the GIF (don't judge me I'm not the best screen recorder :D).

![ezgif-2-727de39f4f](https://user-images.githubusercontent.com/37347167/222540902-b196295a-971f-4c79-9e18-141be9d7ddc9.gif)

Not the prettiest and probably not the best way, but I think it does the job.

I modified a little bit your [Notebook](https://drive.google.com/file/d/1biBRTt-uIuM070oFUNQAaYcDthDQtXsk/view?usp=sharing) so you can have more examples.

**Env:** I used 2xA6000 GPUs from [Paperspace](https://paperspace.com/) using my custom [environent](https://github.com/miguelCalado/docker-paperspace/blob/main/Setup-Info.md)

Let me know how it went on the Kaggle setup @innat!

**Comment by innat** [comment]
@miguelCalado Thanks. Your solution does work. Also, replacing `predict` with `predict_batch` is not required. The only things I need to adopt is to call above `generate_image` within strategy scope (a bit puzzle to me still, as I already invoked the model). Anyway, thanks again.  😄 

```python
with strategy.scope():
    image_gen_model = kcv_models.StableDiffusionV2(
         512, 512, jit_compile=True
    )
    # to warm-up
    _ = generate_image(
            "A"*77, image_gen_model, image_per_prompt=2
        )
```

**Comment by innat** [comment]
Hi @miguelCalado

The above approach you suggseted does make multi-gpu in action but by doing so whether the program is running on one device or both, the execution time is quite different. Here is one demonstration.

```python
# Tesla-T4 (2X)
strategy = tf.distribute.MirroredStrategy(["GPU:0"]) # ["GPU:0", "GPU:1"]
keras.mixed_precision.set_global_policy("mixed_float16")
batch_size = 2
```
```python
def stable_diffusion_model(input_shape):
    with strategy.scope():
        model = kcv_models.StableDiffusionV2(
            *input_shape, jit_compile=True
        )
        
        # to warm-up
        _ = generate_image(
            "A"*77, model, 
            image_per_prompt=strategy.num_replicas_in_sync,
            ug_scale=9
        )
    return model

50/50 [==============================] - 27s 543ms/step
50/50 [==============================] - 27s 534ms/step
50/50 [==============================] - 27s 531ms/step
50/50 [==============================] - 27s 531ms/step
50/50 [==============================] - 27s 534ms/step
```
```python
def stable_diffusion_model(input_shape):
    with strategy.scope():
        model = kcv_models.StableDiffusionV2(
            *input_shape, jit_compile=True
        )
    return model

50/50 [==============================] - 19s 379ms/step
50/50 [==============================] - 19s 378ms/step
50/50 [==============================] - 19s 385ms/step
50/50 [==============================] - 19s 383ms/step
50/50 [==============================] - 19s 381ms/step
```

Full code is same as [above](https://github.com/keras-team/keras-cv/issues/1455#issue-1606095073). 

So, when I tried to warm-up the model and download its weights within `with strategy.scope()`, the execution time (for `batch_size=2`) is increased from `19s` to `27s`.  WDYT?

@LukeWood 
Also, as I said, warming up model inside strategy scope looks anti-pattern, not like usual keras model, doesn't it. It's like calling `model(tf.ones(shape=..))` inside the scope, similar to `model.text_to_image`. About that, what would be the way to run the official [code example](https://keras.io/guides/keras_cv/generate_images_with_stable_diffusion/) on multiple devices?

**Comment by innat** [comment]
@miguelCalado 
Apart from the above situation, here is another observation regarding runtine using variable batch size. For example, with the approach you suggested, if I set `batch_size=10`, the total time takes (using 2 GPUs) around `54` seconds (`5.4` sec each). And with `batch_size=30`, it is `127` second (`4.2` sec each)
